pkg/instrumentation/podmutator.go

```diff
@@ -42,14 +43,142 @@ type instPodMutator struct {
+	instrWithContainers := 0
+	var allContainers []string
+
+	// Check for instrumentations with and without containers.
+	if langInsts.Java.Instrumentation != nil && langInsts.Java.Containers == "" {
+		instrWithoutContainers++
+	} else if langInsts.Java.Instrumentation != nil && langInsts.Java.Containers != "" {
+		allContainers = append(allContainers, langInsts.Java.Containers)
+		instrWithContainers++
+	}
+	if langInsts.NodeJS.Instrumentation != nil && langInsts.NodeJS.Containers == "" {
+		instrWithoutContainers++
+	} else if langInsts.NodeJS.Instrumentation != nil && langInsts.NodeJS.Containers != "" {
+		allContainers = append(allContainers, langInsts.NodeJS.Containers)
+		instrWithContainers++
+	}
+	if langInsts.Python.Instrumentation != nil && langInsts.Python.Containers == "" {
+		instrWithoutContainers++
+	} else if langInsts.Python.Instrumentation != nil && langInsts.Python.Containers != "" {
+		allContainers = append(allContainers, langInsts.Python.Containers)
+		instrWithContainers++
+	}
+	if langInsts.DotNet.Instrumentation != nil && langInsts.DotNet.Containers == "" {
+		instrWithoutContainers++
+	} else if langInsts.DotNet.Instrumentation != nil && langInsts.DotNet.Containers != "" {
+		allContainers = append(allContainers, langInsts.DotNet.Containers)
+		instrWithContainers++
+	}
+	if langInsts.ApacheHttpd.Instrumentation != nil && langInsts.ApacheHttpd.Containers == "" {
+		instrWithoutContainers++
+	} else if langInsts.ApacheHttpd.Instrumentation != nil && langInsts.ApacheHttpd.Containers != "" {
+		allContainers = append(allContainers, langInsts.ApacheHttpd.Containers)
+		instrWithContainers++
+	}
+	if langInsts.Go.Instrumentation != nil && langInsts.Go.Containers == "" {
+		instrWithoutContainers++
+	} else if langInsts.Go.Instrumentation != nil && langInsts.Go.Containers != "" {
+		allContainers = append(allContainers, langInsts.Go.Containers)
+		instrWithContainers++
+	}
+	if langInsts.Sdk.Instrumentation != nil && langInsts.Sdk.Containers == "" {
+		instrWithoutContainers++
+	} else if langInsts.Sdk.Instrumentation != nil && langInsts.Sdk.Containers != "" {
+		allContainers = append(allContainers, langInsts.Sdk.Containers)
+		instrWithContainers++
+	}
+
+	// Look for duplicated containers.
+	containerDuplicates := findDuplicatedContainers(allContainers)
+	if containerDuplicates != nil {
+		return false, fmt.Errorf("duplicated container names detected: %s", containerDuplicates)
+	}
+
+	// Look for mixed multiple instrumentations with and without container names.
+	if instrWithoutContainers > 0 && instrWithContainers > 0 {
+		return false, fmt.Errorf("incorrect instrumentation configuration - please provide container names for all instrumentations")
+		// Look for multiple instrumentations without container names.
+	} else if instrWithoutContainers > 1 && instrWithContainers == 0 {
+		return false, fmt.Errorf("incorrect instrumentation configuration - please provide container names for all instrumentations")
```

Comment:
Is this actually an error?  Isn't this the situation where someone has enabled the feature gate and has a pod with a single container to instrument, such as:

```yaml
apiVersion: apps/v1
kind: Deployment
spec:
  ...
  template:
    metadata:
      annotations:
        instrumentation.opentelemetry.io/inject-dotnet: "true"
```

The above situation should be allowed.

Reply:
Conditions are right. It's `instrWithoutContainers > 1` not `instrWithoutContainers == 1` :) 
The case is allowed - you can see the tests for that: 
- e2e - https://github.com/mat-rumian/opentelemetry-operator/tree/support-multi-cont-instr/tests/e2e-multi-instrumentation/instrumentation-single-instr-first-container
- integration - https://github.com/mat-rumian/opentelemetry-operator/blob/support-multi-cont-instr/pkg/instrumentation/podmutator_test.go#L4378

Reply:
Alright now I'm following, thank you!